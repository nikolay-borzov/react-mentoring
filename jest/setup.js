import promiseFinally from 'promise.prototype.finally'

import Enzyme, { shallow, render, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

// Bring '.finally' to Promise (only for Node 8). Will be available in Node 10
promiseFinally.shim()

Enzyme.configure({ adapter: new Adapter() })

// Make Enzyme functions available in all test files without importing
global.shallow = shallow
global.render = render
global.mount = mount
