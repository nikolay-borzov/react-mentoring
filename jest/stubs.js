export const film = {
  id: 1,
  title: 'Film title',
  poster_path: 'http://via.placeholder.com/500x750',
  overview: 'Film overview',
  release_date: '2018-05-04',
  genres: ['some genre', 'another genre']
}

export const films = [film]
