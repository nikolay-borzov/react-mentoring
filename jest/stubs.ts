import { Film } from '../src/entities/film'

export const film: Film = {
  id: 1,
  title: 'Film title',
  tagline: 'Put that cookie down',
  vote_average: 7.5,
  vote_count: 667,
  release_date: '2018-05-04',
  poster_path: 'https://picsum.photos/g/300/450/?random',
  overview: 'Once upon a time...',
  budget: 15000000,
  revenue: 59735548,
  genres: ['some genre', 'another genre']
}

export const films = [film]

function getRandomImageId() {
  // https://picsum.photos/images
  return Math.floor(Math.random() * Math.floor(1084))
}

export function getRandomFilm(id: number) {
  return {
    ...film,
    id,
    poster_path: film.poster_path.replace(
      'random',
      `image=${getRandomImageId()}`
    )
  }
}
