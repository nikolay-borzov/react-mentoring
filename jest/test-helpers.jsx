import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectly = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectlyShallow = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

/**
 * Creates test closure to check whether component contains another component
 * @param {Function} getComponent Function returning parent component
 * @param {string} componentDisplayName Child component display name
 * @param {Object} params Optional params
 * @param {number} [params.testName=`contains ${componentDisplayName} component`]
 * @param {number} [params.expectedCount=1] Expected number of children components
 * @param {Object} params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent,
  componentDisplayName,
  { testName, expectedCount = 1, expectedProps } = {}
) => {
  it(testName || `contains ${componentDisplayName} component`, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = url => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga, state = {}, arg1 = undefined) => {
  const dispatched = []
  return runSagaOriginal(
    {
      dispatch: action => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
