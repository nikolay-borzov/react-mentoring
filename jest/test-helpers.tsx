import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'
import { AnyAction } from 'redux';

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 */
export const itRendersCorrectly = (
  getComponent: Function,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param getComponent Function returning component to test snapshot
 */
export const itRendersCorrectlyShallow = (
  getComponent: () => React.ReactElement,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

interface ItContainsComponentParams {
  testName?: string
  expectedCount?: number
  expectedProps?: {}
}

/**
 * Creates test closure to check whether component contains another component
 * @param getComponent Function returning parent component
 * @param componentDisplayName Child component display name
 * @param params Optional params
 * @param [params.testName=`contains ${componentDisplayName} component`]
 * @param [params.expectedCount=1] Expected number of children components
 * @param params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent: () => React.ReactElement,
  componentDisplayName: string,
  { testName = `contains ${componentDisplayName} component`, expectedCount = 1, expectedProps }: ItContainsComponentParams = {}
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = (url: string) => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga: any, state = {}, arg1: any = undefined) => {
  const dispatched: AnyAction[] = []
  return runSagaOriginal(
    {
      dispatch: (action: AnyAction) => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
