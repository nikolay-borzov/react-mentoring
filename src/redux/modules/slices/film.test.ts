import { film } from '../../../../jest/stubs'
import { runSaga } from '../../../../jest/test-helpers'

import createSlice, { FilmState } from './film'

import filmService from '../../../services/film-service'

jest.mock('../../../services/film-service')

const initialState: FilmState = {
  item: film,
  error: new Error('Film load error'),
  isFetching: false
}

const parentState = { film: initialState }

const slice = createSlice({
  id: 'FILM',
  rootSelector: state => state.film,
  initialState
})

describe('Film state slice', () => {
  describe('selectors', () => {
    const state = parentState
    const selectors = slice.selectors

    it(`returns 'film'`, () => {
      expect(selectors.film(state)).toBe(state.film.item)
    })

    it(`returns 'error'`, () => {
      expect(selectors.error(state)).toBe(state.film.error)
    })

    it(`returns 'isFetching'`, () => {
      expect(selectors.isFetching(state)).toBe(state.film.isFetching)
    })
  })

  describe('sagas', () => {
    const getFilmMock = filmService.getFilm as jest.Mock

    afterEach(() => {
      getFilmMock.mockClear()
    })

    it('fetches film', () => {
      const response = film

      getFilmMock.mockReturnValue(Promise.resolve(response))

      expect.assertions(2)
      runSaga(slice.sagas.fetchFilmAsync, parentState, {
        payload: film.id
      }).then(dispatched => {
        expect(dispatched).toEqual([
          slice.actionCreators.fetchFilmRequest(),
          slice.actionCreators.fetchFilmSuccess(response)
        ])
        expect(filmService.getFilm).toHaveBeenCalledWith(film.id)
      })
    })

    it('handles film fetch failure', () => {
      const expectedError = new Error('Fetch films failure')

      getFilmMock.mockReturnValue(Promise.reject(expectedError))

      return runSaga(slice.sagas.fetchFilmAsync, initialState, {
        payload: film.id
      }).then(dispatched => {
        expect(dispatched).toEqual([
          slice.actionCreators.fetchFilmRequest(),
          slice.actionCreators.fetchFilmFail(expectedError)
        ])
      })
    })
  })

  describe('reducer', () => {
    const reducer = slice.reducer
    const actionTypes = slice.actionTypes

    it('returns initial state', () => {
      expect(reducer(undefined, { type: 'Unknown' })).toMatchSnapshot()
    })

    it(`handles 'FETCH_FILM_REQUEST'`, () => {
      const state = { isFetching: false }
      const action = {
        type: actionTypes.FETCH_FILM_REQUEST
      }

      expect(reducer(state, action)).toMatchObject({
        isFetching: true
      })
    })

    it(`handles 'FETCH_FILM_SUCCESS'`, () => {
      const state = {
        isFetching: true,
        item: undefined,
        error: new Error('some error')
      }
      const expectedState = {
        isFetching: false,
        item: { id: 1 },
        error: undefined
      }

      const action = {
        type: actionTypes.FETCH_FILM_SUCCESS,
        payload: expectedState.item
      }

      expect(reducer(state, action)).toMatchObject(expectedState)
    })

    it(`it handles 'FETCH_FILM_FAIL'`, () => {
      const state = {
        isFetching: true,
        error: undefined
      }
      const expectedState = {
        isFetching: false,
        error: new Error('some error')
      }
      const action = {
        type: actionTypes.FETCH_FILM_FAIL,
        payload: expectedState.error
      }

      expect(reducer(state, action)).toMatchObject(expectedState)
    })
  })
})
