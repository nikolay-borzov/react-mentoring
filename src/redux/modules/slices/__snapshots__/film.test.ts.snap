// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Film state slice reducer returns initial state 1`] = `
Object {
  "error": [Error: Film load error],
  "isFetching": false,
  "item": Object {
    "budget": 15000000,
    "genres": Array [
      "some genre",
      "another genre",
    ],
    "id": 1,
    "overview": "Once upon a time...",
    "poster_path": "https://picsum.photos/g/300/450/?random",
    "release_date": "2018-05-04",
    "revenue": 59735548,
    "tagline": "Put that cookie down",
    "title": "Film title",
    "vote_average": 7.5,
    "vote_count": 667,
  },
}
`;
